### DISCUSSION TITLE
Make a field frozen during init

### DISCUSSION THREAD

**Discussion by orfisko** [pydantic/pydantic#7100] [Question]
Might be a silly question, but does someone know how to freeze a field during init? I can also simply make a different childclass with this field frozen, but I am just wondering whether this is actually possible.

```python
from pydantic import BaseModel


class Test(BaseModel):
    foo: int

    def __init__(self, frozen: bool = False, **data):
        if frozen:
            self.model_fields["foo"].frozen = frozen #This does not work
        super().__init__(**data)

t1 = Test(foo=1, frozen=True)
t1.foo = 4 # Works, but would love it to fail
print(t1) # foo=4

**Comment by Viicos** [comment]
See https://github.com/pydantic/pydantic/pull/7103, you will probably be able to pull it off if accepted. Otherwise it might be tricky, as `__getattr__` relies on things set on `pydantic-core`, which isn't easy to patch after model creation

**Comment by orfisko** [comment]
@Viicos thanks for your reply. My test also fails when i have validate_assignment switched to True.

**Reply by Viicos** [reply]
The example you gave will work once the PR is merged. Note that `frozen` is the only field that will be taken into account when tweaking `model_fields` in the `__init__` method (as it will be looked up in `__setattr__`. 

But anyway, it isn't a good idea to do it that way:

```python
class Test(BaseModel):
    foo: int

    def __init__(self, frozen: bool = False, **data):
        if frozen:
            self.model_fields["foo"].frozen = frozen #This does not work
        super().__init__(**data)

Test.model_fields
#> {'foo': FieldInfo(annotation=int, required=True)}
t1 = Test(foo=1, frozen=True)
Test.model_fields
#> {'foo': FieldInfo(annotation=int, required=True, frozen=True)}
# The change has been made to the class, and not only the instance!!!
```

`model_fields` is a `ClassVar`, meaning it shouldn't be modified from an instance (type checkers will raise an error with this code).

I'm not aware of any mechanism to provide customization at instanciation, I think it would be easier for you to make some kind of intermediary model factory.
